### ISSUE TITLE
is it possible to build NARX network using this library?

### ISSUE THREAD

**Issue by danil-z** [cazala/synaptic#219]
<img width="594" alt="screen shot 2017-05-02 at 8 32 14 am" src="https://cloud.githubusercontent.com/assets/509032/25626513/9d5e4830-2f14-11e7-989d-75efdc6bcafe.png">

**Comment by wagenaartje** [comment]
Never heard of it before, did some research, basically works like this:

![img](https://i.gyazo.com/da10b2fe9d03486af2e48aade605c254.png)

So to predict the new value of `y`, previous inputs and outputs are used as input, right? I'm also seeing some different examples using only previous outputs as input.  But I think it is possible in this library, you have to use recurrent connections though.

**Comment by clark-nextoner** [comment]
Curious what you find here

**Comment by wagenaartje** [comment]
So I worked it out and assuming then etwork takes input u<sub>t</sub> to u<sub>t-n</sub> and output y<sub>t-1</sub> to y<sub>t-n</sub> as input, it's definitely possible. Schematic:

![](http://i.imgur.com/qcLyVcw.png)

So basically to copy the previous inputs to the next node, we need connections with a constant weight (`=1`) and nodes with a bias of `0` and the `IDENTITY` function as activation. Every activation sequence the previous activation will be moved to the next node. The same is done for outputs.

The previous outputs are fed into the hidden layer, together with the current input and previous inputs.

The only change this requires in synaptic codes is to make some nodes non-propogatable: aka they should not modify their bias and inputting connection weights. This seems do-able. 

<hr>

TL;DR this can be implemented. I'll start working on NARX networks, all I need is some dataset that NARX networks should be able to achieve low error on (versus regular networks).

**Comment by wagenaartje** [comment]
Update: I managed to build a NARX network. However, I have doubts in the backpropagation algorithm. Does anyone have a training set?

**Comment by wagenaartje** [comment]
So I implemented it. Check it out:

* [Wiki](https://github.com/wagenaartje/neataptic/wiki/Architect#narx)
* [JSFiddle](https://jsfiddle.net/wagenaartje/1o7t91yk/)
* [Code](https://github.com/wagenaartje/neataptic/blob/master/src/architect.js#L276)

Let me know if you have any problems.

**Comment by cazala** [comment]
This sounds like the context neuron in Elman networks (neurons holding previous activations with a fixed connection weight of 1). Those type of networks were outplaced by RNNs which are contemplated in the LSTM-g algorithm. When a neuron is self-connected, it has a fixed weight `Wjj=1`. You can do self connections in synaptic like the this:

```
var inputLayer = new Layer(3);
var hiddenLayer = new Layer(10);
var outputLayer = new Layer(2);

inputLayer.project(hiddenLayer);
hiddenLayer.project(outputLayer);

hiddenLayer.project(hiddenLayer); // recurrent connection

var RNN = new Network({
input: inputLayer,
hidden: [hiddenLayer]
output: outputLayer
});
```
